Regression

jasonkyuyim opened the pull request:
This PR adds regression functionality to RerF.

Changes:
* Split BuildTree into classTree and regTree.  classTree gets built when classification is specified, regTree when regression. 
* Added regression parameter to RerF main function
* Added regression split function to split.cpp.
* Functions now have a `task` parameter to specify whether classification or regression is being performed.
* Formatting was beautified in several function headers and file comments, i.e. `split.cpp` has file comments, parameters are separated on new lines, tab sizes were inconsistent (sometimes 2,4, or even 8). 

**NOTE: I couldn't add tests RerF doesn't use testthat and travis pulls from the neurodatamaster. I developed on my own fork I was unable to add tests and have it be refelcted in travis. I will open a new branch in neurodata/master that has regression tests once this gets accepted and passes the current tests.**

MrAE commented:
I'm looking through this.  We modified travisCI so that tests will run on all branches and PR's. 
So @jasonkyuyim, you can should now be able to add tests.

falkben commented:
This pull request looks good, thank you!  

Can you please add tests for this new feature to the travisTest folder?

jovo commented:
@jasonkyuyim status?

jasonkyuyim commented:
Sorry for not updating this. I have a sprint happening at work that requires my time right now. The tests  should be very easy to make. If someone wants to take it over I can send over a driver that runs regression and then making a testthat out of it will be easy. Maybe a day's work. Otherwise I will resume and finish in a week when I have time.

jovo commented:
please send driver either way.
thanks.

falkben commented on R/Predict.R:
```diff
@@ -75,24 +75,32 @@ Predict <-
 
     if (!aggregate.output) {
-      predictions <- matrix(forest$labels[unlist(Yhats)], nrow(X), f_size)
+      if (task == 'classificaiton')
```
Spelling mistake which means classification predictions are never run here

jasonkyuyim replied:
Fixed.

falkben commented on R/Predict.R:
```diff
@@ -75,24 +75,32 @@ Predict <-
+            predictions <- factor(forest$labels[max.col(predictions)], levels = forest$labels)
+          }
         }
+      } else {
```
we probably want to change this to else if to be clear about what's going on

jasonkyuyim replied:
I added a check at that beginning of the function to check the task is either classification or regression. The design of the if-else statements is very bad. Changing every one of them to be tryCatch is not good design. It should be abstracted into a single class and methods should be overloaded based on whether regression or classification is used.

falkben commented on R/classTree.R:
```diff
@@ -21,7 +21,7 @@
 #'
 
-BuildTree <-
+classTree <-
```
Do we need a rename here?

jasonkyuyim replied:
it was easier this way to develop at the time. Random Forest in CRAN and Breiman's original implementation have classTree and regTree as classes for building classification and regression trees. Now, the two files share a lot of similarities so it would make much more sense to have base tree building class but I didn't have time to do that. Otherwise we risk muddying up a single function with a bunch of switch statements and debugging/developing on that is a nightmare.

falkben commented on src/split.cpp:
```diff
@@ -118,3 +136,54 @@ List findSplit(const NumericVector x, const IntegerVector y, const int & ndSize,
+	       double maxdI, int bv, double bs, const int nzidx, arma::vec cc, const int & task) {
+
+    List ret;
+    if (task == 0) {
```
can we be more explicit about what we are doing here?

jasonkyuyim replied:
Added comments.

jasonkyuyim commented:
Addressed comments. Sent the script/driver to Jesse. There's clearly a lot of refactoring that could be in these files to make it more modular and cleaner such as making base classes and only overriding certain functions for classification vs. regression. It's difficult to add type checking and diversify the functionality without making the files huge and difficult to debug/read.
